### PR TITLE
`Programming exercises`: Add typescript programming language template

### PR DESCRIPTION
**Pull request by magaupp**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).

#### Client
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).
- [x] I tested **all** changes and their related features with **all** corresponding user types on a _local_ server configured with **Jenkins**.

### Description

This PR adds a TypeScript programming exercise template. The exercise is very similar to the JavaScript template.

The test project builds the exercise as a composite with `tsc` and runs the tests with `ts-jest`.

Compared to JS, this exercise adds type annotations, the SortStrategy as a interface and Generics.
The tests cast the student implementations to `any`, which prevents build failures. The resulting runtime errors produce proper test failures.

The JavaScript Docker image is reused without changes.

### Steps for Testing

Prerequisites:
- 1 Course

1. Navigate to the exercise creation page
2. Select `TypeScript` as the programming language
3. Fill out the rest of the required fields. Keep the default build plan
4. Generate the exercise
5. Wait until the `Template` and `Solution` build jobs finish
6. The `Template Result` should show 0 out of 10 passed tests
7. The `Solution Result` should show 10 out of 10 passed tests

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review
#### Manual Tests
- [x] Test

## Summary by CodeRabbit

- **New Features**
	- Added support for TypeScript as a programming language within the application.
	- Introduced TypeScript-specific build and testing configurations.
	- Implemented sorting algorithms (BubbleSort and MergeSort) using the Strategy Pattern in TypeScript.
	- Enhanced the user documentation to include TypeScript capabilities.
	- Added a new Bash script for automating TypeScript project setup and testing.

- **Bug Fixes**
	- Improved handling of programming language features in Jenkins and local CI environments.

- **Tests**
	- Added unit tests for sorting algorithms and policy-based sorting strategy.

- **Documentation**
	- Updated user documentation to reflect new TypeScript support and features.

### CODE REVIEW

File: src/main/resources/templates/typescript/exercise/src/mergesort.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export default class MergeSort {
+    // TODO: implement in performSort(Array<Date>)
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/exercise/src/mergesort.ts:2]: All files under an `exercise` directory contain exercises where TODO comments mark tasks solved by students.

File: src/main/resources/templates/typescript/exercise/src/mergesort.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
...
+}
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/exercise/src/mergesort.ts:3]: All files under an `exercise` directory contain exercises where TODO comments mark tasks solved by students.
See the `solution` directory for the proper implementation.

File: src/main/resources/templates/typescript/exercise/src/bubblesort.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export default class BubbleSort {
+    // TODO: implement in performSort(Array<Date>)
+}
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/exercise/src/bubblesort.ts:3]: The class structure is intentionally left as an exercise.

File: src/main/resources/templates/typescript/exercise/src/context.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+export default class Context {
+    // TODO: Create and implement a Context class according to the UML class diagram
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/exercise/src/context.ts:2]: The `readme` contains a detailed task description.

File: src/main/resources/templates/typescript/exercise/src/policy.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+export default class Policy {
+    // TODO: Create and implement a Policy class as described in the problem statement
+}
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/exercise/src/policy.ts:3]: The class structure is intentionally left as an exercise.
A showcase of best practices should be under the `solution` directory.

File: src/main/resources/templates/typescript/exercise/src/sortstrategy.ts [hunk 6]
```diff
@@ -0,0 +1,3 @@
+export default interface SortStrategy {
+    // TODO: Create a SortStrategy interface according to the UML class diagram
+}
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/exercise/src/sortstrategy.ts:3]: The class structure is intentionally left as an exercise.

File: src/main/resources/templates/jenkins/typescript/regularRuns/pipeline.groovy [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    stage('Install Dependencies') {
+        sh 'npm ci --prefer-offline --no-audit'
+    }
+    stage('Build') {
+        sh 'npm run build'
+    }
+    stage('Test') {
+        sh 'npm run test:ci'
+    }
+}
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/jenkins/typescript/regularRuns/pipeline.groovy:40]: This only builds the assignment as a composite sub-project. The root project has `noEmit` set.

File: src/main/resources/templates/typescript/exercise/package.json [hunk 8]
```diff
@@ -0,0 +1,19 @@
+{
+    "name": "artemis-exercise",
+    "private": true,
+    "scripts": {
+        "build": "tsc",
+        "start": "node ./dist/client.js"
+    },
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/exercise/package.json:7]: The tests are not run by the students.

File: src/main/resources/templates/typescript/solution/package.json [hunk 9]
```diff
@@ -0,0 +1,19 @@
+{
+    "name": "artemis-exercise",
+    "private": true,
+    "scripts": {
+        "build": "tsc",
+        "start": "node ./dist/client.js"
+    },
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/solution/package.json:7]: The tests are not run by the students.

File: src/main/resources/templates/typescript/solution/package.json [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+    "exports": {
+        "./*": {
+            "types": "./dist/*.d.ts",
+            "default": "./dist/*.js"
+        }
+    },
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/solution/package.json:13]: This is not a real library. A wildcard makes testing this package the easiest.

File: src/main/resources/templates/aeolus/typescript/default.sh [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+  fi
+  local _script_name
+  _script_name=${BASH_SOURCE[0]:-$0}
+  cd "${AEOLUS_INITIAL_DIRECTORY}"
+  bash -c "source ${_script_name} aeolus_sourcing; install_dependencies"
+  cd "${AEOLUS_INITIAL_DIRECTORY}"
+  bash -c "source ${_script_name} aeolus_sourcing; build"
+  cd "${AEOLUS_INITIAL_DIRECTORY}"
+  bash -c "source ${_script_name} aeolus_sourcing; test"
+}
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/aeolus/typescript/default.sh:31]: Code in `.sh` files under the `aeolus` directory is usually generated code.

File: src/main/resources/templates/typescript/exercise/src/client.ts [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+        console.log('Unsorted Array of dates:');
+        console.log(dates);
+
+        // TODO: Sort dates
+
+        console.log('Sorted Array of dates:');
+        console.log(dates);
+    }
+}
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/exercise/src/client.ts:29]: Code under the `exercise` directory is a template where TODO comments mark tasks to be implemented by students.

File: src/main/resources/templates/typescript/exercise/src/client.ts [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * @param low {number} the lower bound
+ * @param high {number} the upper bound
+ * @returns {number} random int within the given range
+ */
+function randomIntegerWithin(low: number, high: number): number {
+    return Math.floor(Math.random() * (high - low + 1)) + low;
+}
+
+main();
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/exercise/src/client.ts:69]: The upper bound should be inclusive

File: src/main/resources/templates/typescript/solution/src/bubblesort.ts [hunk 14]
```diff
@@ -0,0 +1,21 @@
+import SortStrategy from './sortstrategy';
+import Comparable from './comparable';
+
+export default class BubbleSort<T extends Comparable = Date> implements SortStrategy<T> {
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/solution/src/bubblesort.ts:4]: The default type is required for the tests to work with generic and non-generic implementations

File: src/main/resources/templates/typescript/test/package.json [hunk 15]
```diff
@@ -0,0 +1,24 @@
+{
+    "name": "artemis-test",
+    "private": true,
+    "scripts": {
+        "build": "tsc -b",
+        "test": "jest",
+        "test:ci": "jest --ci --reporters=default --reporters=jest-junit"
+    },
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/test/package.json:8]: The test package exists only for testing

File: src/main/resources/templates/typescript/test/package.json [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+    "workspaces": [
+        "assignment"
+    ],
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/test/package.json:11]: The assignment folder is copied in by the CI beforehand

File: src/main/resources/templates/typescript/readme [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+Add `get` and `set` accessors for the attribute.
+`Policy` should implement a simple configuration mechanism:
+
+    1. [task][Select MergeSort](behavior_Policy_uses_MergeSort_for_big_list)
+    Select `MergeSort` when the List has more than 10 dates.
+
+    2. [task][Select BubbleSort](behavior_Policy_uses_BubbleSort_for_small_list)
+    Select `BubbleSort` when the List has less or equal 10 dates.
+
+4. Complete the `main()` function which demonstrates switching between two strategies at runtime.
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/readme:42]: The provided main() function includes sufficient instructions in comments

File: src/main/resources/templates/typescript/solution/src/context.ts [hunk 18]
```diff
@@ -0,0 +1,30 @@
+import type SortStrategy from './sortstrategy';
+
+export default class Context {
+    private _sortAlgorithm: SortStrategy | null = null;
+
+    private _dates: Array<Date> = [];
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/solution/src/context.ts:6]: dates has a setter

File: src/main/resources/templates/typescript/solution/src/context.ts [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+        this._sortAlgorithm?.performSort(this._dates);
+    }
+
+    get sortAlgorithm(): SortStrategy | null {
+        return this._sortAlgorithm;
+    }
+
+    set sortAlgorithm(sortAlgorithm: SortStrategy) {
+        this._sortAlgorithm = sortAlgorithm;
+    }
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/solution/src/context.ts:21]: Setting null is not useful here

File: src/main/resources/templates/typescript/solution/src/policy.ts [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+    /**
+     * Chooses a strategy depending on the number of date objects.
+     */
+    configure() {
+        if (this._context.dates.length > DATES_LENGTH_THRESHOLD) {
+            this._context.sortAlgorithm = new MergeSort();
+        } else {
+            this._context.sortAlgorithm = new BubbleSort();
+        }
+    }
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/solution/src/policy.ts:19]: _context by its type cannot be null or undefined

File: src/main/resources/templates/typescript/solution/src/policy.ts [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+
+    get context(): Context {
+        return this._context;
+    }
+
+    set context(context: Context) {
+        this._context = context;
+    }
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/solution/src/policy.ts:27]: The type system checks this

File: src/main/resources/templates/typescript/solution/src/policy.ts [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+}
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/solution/src/policy.ts:28]: Tests are implemented in the `test` directory.

File: src/main/resources/templates/typescript/solution/src/bubblesort.ts [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+        for (let i = input.length - 1; i >= 0; i--) {
+            for (let j = 0; j < i; j++) {
+                if (input[j].valueOf() > input[j + 1].valueOf()) {
+                    const temp = input[j];
+                    input[j] = input[j + 1];
+                    input[j + 1] = temp;
+                }
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/solution/src/bubblesort.ts:20]: We want the built-in Date objects to work out of the box.

File: src/main/resources/templates/typescript/solution/src/mergesort.ts [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+/**
+ * Recursive merge sort function
+ *
+ * @template T
+ * @param input {Array<T>}
+ * @param low {number}
+ * @param high {number}
+ */
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/solution/src/mergesort.ts:23]: IDEs show an error for an unresolved type T if it isn't declared

File: src/main/resources/templates/typescript/solution/src/mergesort.ts [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ * Merge function
+ *
+ * @template T
+ * @param input {Array<T>}
+ * @param low {number}
+ * @param middle {number}
+ * @param high {number}
+ */
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/solution/src/mergesort.ts:42]: IDEs show an error for an unresolved type T if it isn't declared

File: src/main/resources/templates/typescript/solution/src/mergesort.ts [hunk 26]
```diff
@@ -0,0 +1,69 @@
+import SortStrategy from './sortstrategy';
+import Comparable from './comparable';
+
+export default class MergeSort<T extends Comparable> implements SortStrategy<T> {
+    /**
+     * Wrapper method for the real MergeSort algorithm.
+     *
+     * @template T
+     * @param input {Array<T>} the array of objects to be sorted
+     */
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/solution/src/mergesort.ts:10]: IDEs show an error for an unresolved type T if it isn't declared

File: src/main/resources/templates/typescript/exercise/tsconfig.json [hunk 27]
```diff
@@ -0,0 +1,8 @@
+{
+  "extends": "@tsconfig/node20/tsconfig.json",
+  "compilerOptions": {
+    "outDir": "./dist",
+    "rootDir": "./src",
+    "composite": true
+  }
+}
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/typescript/exercise/tsconfig.json:8]: Instructors can decide to set a stricter configuration.

**Review by SimonEntholzer — APPROVED** [review]
Works nicely on Ts2

**Review by HawKhiem — APPROVED** [review]
Tested on TS3. Works like described 
Screenshot 2024-10-16 203414
Screenshot 2024-10-16 203601

**Review by Cathy0123456789 — APPROVED** [review]
Tested on TS1. Works as expected.

**Review by flbrgit — COMMENTED** [review]
On TS4 no errors occured, everything works smoothly and as described
Create _ Practical Course_ Interactive Learning WS24_25 - Vivaldi 20 10 2024 12_18_50

**Review by SindiBuklaji — APPROVED** [review]
Tested on TS3. It worked good without errors. 10/10 tests passed for the solution repo and 0/10 passed for the template repo.

**Review by SimonEntholzer — APPROVED** [review]
Code looks good to me :+1:

**Pull request merged**